File: lib/py/uapi/internal/generation/GenerateRandomObject.py
from typing import TYPE_CHECKING, cast

if TYPE_CHECKING:
    from uapi.RandomGenerator import RandomGenerator
    from uapi.internal.types.UTypeDeclaration import UTypeDeclaration
    from uapi.internal.generation.GenerateContext import GenerateContext


def generate_random_object(blueprint_value: object, use_blueprint_value: bool, type_parameters: list['UTypeDeclaration'], ctx: 'GenerateContext') -> object:
    nested_type_declaration = type_parameters[0]

    if use_blueprint_value:
        starting_obj = cast(dict[str, object], blueprint_value)

        obj = {}
        for key, starting_obj_value in starting_obj.items():
            value = nested_type_declaration.generate_random_value(
                starting_obj_value, True, ctx)
            obj[key] = value

        return obj
    else:
        length = ctx.random_generator.next_collection_length()

        obj = {}
        for i in range(length):
            key = ctx.random_generator.next_string()
            value = nested_type_declaration.generate_random_value(
                None, False, ctx)
            obj[key] = value

        return obj

File: lib/java/src/main/java/uapi/internal/generation/GenerateRandomObject.java
package uapi.internal.generation;

import java.util.List;
import java.util.Map;
import java.util.TreeMap;

import uapi.internal.types.UTypeDeclaration;

public class GenerateRandomObject {
    public static Object generateRandomObject(Object blueprintValue, boolean useBlueprintValue,
            List<UTypeDeclaration> typeParameters, GenerateContext ctx) {
        final var nestedTypeDeclaration = typeParameters.get(0);

        if (useBlueprintValue) {
            final var startingObj = (Map<String, Object>) blueprintValue;

            final var obj = new TreeMap<String, Object>();
            for (final var startingObjEntry : startingObj.entrySet()) {
                final var key = startingObjEntry.getKey();
                final var startingObjValue = startingObjEntry.getValue();
                final var value = nestedTypeDeclaration
                        .generateRandomValue(startingObjValue, useBlueprintValue, ctx);
                obj.put(key, value);
            }

            return obj;
        } else {
            final var length = ctx.randomGenerator.nextCollectionLength();

            final var obj = new TreeMap<String, Object>();
            for (int i = 0; i < length; i += 1) {
                final var key = ctx.randomGenerator.nextString();
                final var value = nestedTypeDeclaration.generateRandomValue(null, false, ctx);
                obj.put(key, value);
            }

            return obj;
        }
    }
}
